Content Card Developer Use Case Doc (#1475)
Appboy/braze-docs: _docs/_developer_guide/platform_integration_guides/ios/content_cards/best_practices.md

---

# Best Practices and Use Cases
{% include video.html id="e0G7IJfkMOI" align="right" %}

> This developer article covers Content Card best practices, three use cases built by our team, accompanying code snippets, as well as an implementation walkthrough. Included to the right is our full video walkthrough, with sections included and referenced throughout the rest of the article.

## Best Practices

### Code Considerations

When adding Content Cards to your codebase, it is a best practice to decouple your code. By keeping your code separate from the SDK code, it makes it easier to troubleshoot. For example, try to minimize your SDK imports down to a single "import Appboy-iOS-SDK" statement. This limits issues that arise from excessive SDK imports, making it easier to track, debug, and alter code.
- __Decoupled__ - Your data should be source-independent. Decoupled code allows you to easily track where your Content Card data is coming from and going to.
- __Flexibility__ - Offers type-agnostic extended functionality. You can extend your custom production code to handle Content Card data.

> - __Easy to Debug__ - Errors can be traced to one place. Because of the minimization of SDK imports, your code now has a single source of data to check and debug along. 

Reviewer: ```suggestion
- __Easy to Debug__ - Errors can be traced to one place. Because of the minimization of SDK imports, your code now has a single source of data to check and follow along. 
```

## Sample Use Cases

> There are three customer use cases provided. We have provided video, code snippets for each demo application, as well as dashboard implementation examples for each. The use cases are:

Reviewer: Maybe something along the lines of:
```suggestion
There are 3 sample customer use cases provided. Each sample has video walkthroughs, code snippets for each demo application, and dashboard implementation examples:
```

#### Load Content Cards Alongside Existing Content
Get the data simultaneously through Operation Queues
```swift
addOperation { [weak self] in
      guard let self = self else { return }
      self.loadCourses(self.courseCompletionHandler)
    }
    
addOperation { [weak self] in
      guard let self = self else { return }
      self.loadContentCards()
      self.semaphore.wait()
    }
```
Course Operation
```swift
func loadCourses(_ completion: @escaping ([Course]) -> ()) {
      switch result {
      case .success(let metaData):
        completion(metaData.courses)
      case .failure:
        cancelAllOperations()
    }
  }
```
Content Card Operation
```swift
func loadContentCards() {
    AppboyManager.shared.addObserverForContentCards(observer: self, selector: #selector(contentCardsUpdated))
    AppboyManager.shared.requestContentCardsRefresh()
  }

@objc private func contentCardsUpdated(_ notification: Notification) {
    let contentCards = AppboyManager.shared.handleContentCardsUpdated(notification, for: [.item(.course), .ad])
    contentCardCompletionHandler(contentCards)
    semaphore.signal()
  }
```

#### Extending Course Functionality

> Populating Course Object through Content Card Payload

Reviewer: ```suggestion
Populating Course Object with the `Content Card` Payload
```

Also, is it intentional to capitalize most of these words here - like Course, Object, Payload?

```swift
// MARK: - Course
struct Course: ContentCardable, Purchasable, Codable, Hashable {
  private(set) var contentCardData: ContentCardData?
  let id: Int
  let title: String
  let price: Decimal
  let imageUrl: String
    
  private enum CodingKeys: String, CodingKey {
    case id
    case title
    case price
    case imageUrl = "image"
  }
}
```

#### ABKContentCard Dependencies

> The Only Dependencies on ABKContentCard are its Primitive Types

Reviewer: ```suggestion
The Only Dependencies on `ABKContentCard` are its Primitive Types
```

```swift
protocol ContentCardable {
  var contentCardData: ContentCardData? { get }
  init?(metaData: [ContentCardKey: Any], classType contentCardClassType: ContentCardClassType)
}

extension ContentCardable {
  var isContentCard: Bool {
    return contentCardData != nil
  }

// MARK: - ContentCardData
struct ContentCardData: Hashable {
  let contentCardId: String
  let contentCardClassType: ContentCardClassType
  let createdAt: Double
  let isDismissable: Bool
}
```

#### Campaign Key-Value Pairs From Braze Dashboard
```swift

> // MARK: - Content Card Initalizer

Reviewer: I don't think it'll be necessary to have the `MARK`'s in public docs

extension Course {
  init?(metaData: [ContentCardKey: Any], classType contentCardClassType: ContentCardClassType) {
    guard let idString = metaData[.idString] as? String,
      let createdAt = metaData[.created] as? Double,
      let isDismissable = metaData[.dismissable] as? Bool,
      let extras = metaData[.extras] as? [AnyHashable: Any],
      let title  = extras["course_title"] as? String,
      let detail = extras["course_detail"] as? String,
      let priceString = extras["course_price"] as? String,
      let price = Decimal(string: priceString),
      let imageUrl = extras["course_image"] as? String
      else { return nil }
```

#### Identifying Multiple Types
```swift
init(rawType: String?) {
    switch rawType?.lowercased() {
    case "ad":
      self = .ad
    case "coupon_code":
      self = .coupon
    case "course_recommendation":
      self = .item(.course)
    case "message_classic":
      self = .message(.fullPage)
    case "message_webview":
      self = .message(.webView)
    default:
      self = .none
    }
  }
```
{% endtab %}
{% endtabs %}

### Content Cards in a Message Center

> Content Cards can be used in a message center format where each message is its own Content Card. Each Content Card contains additional key-value pairs that power on-click UI/UX.

Reviewer: ```suggestion
Content Cards can be used in a message center format where each message is its own card. Each one can contain additional key-value pairs that power on-click UI/UX.
```

{% tabs %}
{% tab Video %}

{% include video.html id="ZpXvjca9KyY" align="center" %}

{% endtab %}
{% tab Swift - Demo Code Snippets %}
#### Using Class_type for On Click Behavior
How to Filter and Identify Various Class Types
```swift
func addContentCardToView(with message: Message) {
    switch message.contentCardData?.contentCardClassType {
    case .message(.fullPage):
      loadContentCardFullPageView(with: message as! FullPageMessage)
    case .message(.webView):
      loadContentCardWebView(with: message as! WebViewMessage)
    default:
      break
    }
}
```
{% endtab %}
{% endtabs %}

### Interactive Content Cards

> Content Cards can be leveraged to create interactive experiences for your users. In the demo below, we have a Content Card pop-up appear at checkout providing users last-minute promotions. Well placed cards like this are a great way to give users a "nudge" toward specific user actions. 

Reviewer: ```suggestion
Content Cards can be leveraged to create interactive experiences for your users. In the demo below, we have a Content Card pop-up appear at checkout providing users last-minute promotions. Well-placed cards like this are a great way to give users a "nudge" toward specific user actions. 
```
